docs_nnx/guides/pytree.md

```diff
@@ -196,64 +306,119 @@ foo = Foo()
+    self.right = self.left  # reference sharing
+
+m = Parent()
+
+print(f"{nnx.find_duplicates(m) = }  # not a tree")
+```
+
+The main issue is that sharing is not preserved across pytree operations including JAX transforms, and this results in unintended state duplication:
+
+```{code-cell} ipython3
+m = Parent()
+print(f"Before: {m.left is m.right = }")
+
+@jax.jit
+def f(m):
+  print(f"Inside: {m.left is m.right = }")
+  return m
+
+m = f(m)
+print(f"After:  {m.left is m.right = }")
+```
+
+Reference sharing is rare in most Machine Learning applications, however if it is required you can either use the `nnx.{split, merge, state, update}` APIs to move the deduplicated state and graph definiton across the JAX transforms:
+
+```{code-cell} ipython3
+m = Parent()
+print(f"Before: {m.left is m.right = }")
+graphdef, state = nnx.split(m)
+
+@jax.jit
+def f(graphdef, state):
+  m = nnx.merge(graphdef, state)
+  print(f"Inside: {m.left is m.right = }")
+  return nnx.state(m)
+
+state = f(graphdef, state)
+nnx.update(m, state)
+
+print(f"After:  {m.left is m.right = }")
+print(f"{state = }") # deduplicated state
 ```
 
+Or alternatively you can use the NNX transforms which preserve shared references:
+
 ```{code-cell} ipython3
-model = Block(din=1, dout=2, rngs=nnx.Rngs(0))
-
-model.eval(mode=2)  # .set_attributes(deterministic=True, use_running_average=True, mode=2)
-print("eval:")
-print(f"- {model.mode = }")
-print(f"- {model.bn.use_running_average = }")
-print(f"- {model.dropout.deterministic = }")
-
-model.train(mode=1)  # .set_attributes(deterministic=False, use_running_average=False, mode=1)
-print("train:")
-print(f"- {model.mode = }")
-print(f"- {model.bn.use_running_average = }")
-print(f"- {model.dropout.deterministic = }")
+m = Parent()
+print(f"Before: {m.left is m.right = }")
+
+@nnx.jit
+def f(m):
+  print(f"Inside: {m.left is m.right = }")
+  return m
+
+m = f(m)
+
+print(f"After:  {m.left is m.right = }")
+```
+
+### pytree=False
+`nnx.Pytree` allows you to turn off the pytree registration along with the attribute checks for subtypes by setting `pytree` type attribute option to `False`. This can be useful when upgrading to previous NNX code to newer Flax verions as you will still be able to use the NNX APIs or when creating types that should not be treated as pytree because e.g. they shared references.
+
+```{code-cell} ipython3
+class Foo(nnx.Pytree, pytree=False):
+  def __init__(self):
+    self.a = [jnp.array(1), jnp.array(2)]  # no checks
+    self.b = "hello" 
+    self.b = jnp.array(3) # no checks
+
+foo = Foo()
+
+@nnx.jit # can use in NNX transformations
+def double(foo: Foo):
+  foo.a = [x * 2 for x in foo.a]
+  foo.b *= 2
+
+double(foo)
+print(f"{ nnx.state(foo) = }")  # can be used with NNX APIs
+print(f"{ jax.tree_util.all_leaves([foo]) = }")  # not a pytree
 ```
 
+## Module
+
++++
+
+NNX Modules are `Pytree`s that have two additional methods for traking intermediate values: `sow` and `perturb`.
+
++++
+
 ### sow
```

Comment:
Should we move sow and perturb into a separate guide? This will make it easier when we move to boxes and capture context.

Reply:
Would be a good fit for a Capture Intermediates guide. Adding it here for now.